dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,42 +52,203 @@ public Pattern getParameterPattern() {
-            int position = 0;
-            while (rs.next()) {
-                String columnName = rs.getString(1);
-                String dataType = rs.getString(2);
-                DbParameterAccessor dbp = createDbParameterAccessor(
-                        columnName,
-                        Direction.INPUT,
-                        typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = getColumnName(rs);
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, columnDirection(rs),
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+
+        protected String getColumnName(ResultSet rs) throws SQLException {
+            return defaultIfNull(rs.getString("COLUMN_NAME"), "");
+        }
+
+        protected Direction columnDirection(ResultSet rs) throws SQLException {
+            Direction columnDir = getColumnDirection(rs);
+            if (columnDir == null) {
+                throw new SQLException("Invalid type for column/parameter " + getColumnName(rs));
+            }
+            return columnDir;
+        }
+
+        abstract protected Direction getColumnDirection(ResultSet rs) throws SQLException;
+    }
+
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getColumnDirection(ResultSet rs) {
+            return Direction.INPUT;
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        abstract protected Direction getColumnDirection(ResultSet rs) throws SQLException;
```

Comment:
Abstract stuff doesn't need to be overridden (it's inherited). But here `DatabaseExecutable` is already completely hollow so it can be dropped.